rag_chat/conversation_context.py

```diff
@@ -0,0 +1,304 @@
+        # Split into recent (keep verbatim) and older (summarize) sections
+        recent_messages = history[-self.config.max_recent_messages:]
+        older_messages = history[:-self.config.max_recent_messages]
+        
+        # Summarize the older messages
+        summary = self._create_conversation_summary(older_messages, session)
+        
+        # Build final conversation with summary + recent + current
+        result = []
+        
+        if summary:
+            result.append({"role": "system", "content": f"Previous conversation summary: {summary}"})
+        
+        result.extend(recent_messages)
+        result.append(current_message)
+        
+        # Check if we're still within limits, truncate recent if needed
+        total_tokens = sum(self._count_tokens(msg["content"]) for msg in result)
+        if total_tokens > self.config.max_conversation_tokens:
+            # Need to truncate recent messages too
+            return self._truncate_conversation(recent_messages + [current_message])
+        
+        return result
+    
+    def _hybrid_conversation(self, conversation: List[Dict[str, str]], session: ChatSession) -> List[Dict[str, str]]:
+        """
+        Hybrid strategy - use summarization when beneficial, truncation otherwise.
+        """
+        current_message = conversation[-1]
+        history = conversation[:-1]
+        
+        # Calculate tokens in history
+        history_tokens = sum(self._count_tokens(msg["content"]) for msg in history)
+        
+        # If history is under summarization threshold, just truncate
+        if history_tokens < self.config.summarization_threshold:
+            return self._truncate_conversation(conversation)
+        
+        # If we have enough messages and tokens to make summarization worthwhile
+        if len(history) >= self.config.max_recent_messages * 2:
+            return self._summarize_conversation(conversation, session)
+        
+        # Fall back to truncation for edge cases
+        return self._truncate_conversation(conversation)
+    
+    def _create_conversation_summary(self, messages: List[Dict[str, str]], session: ChatSession) -> Optional[str]:
+        """
+        Create a concise summary of conversation messages using LLM.
+        
+        Args:
+            messages: List of conversation messages to summarize
+            session: Chat session for context
+            
+        Returns:
+            Summary text or None if summarization fails
+        """
+        if not messages:
+            return None
+        
+        try:
+            # Build the conversation text to summarize
+            conversation_text = "\n\n".join([
+                f"{msg['role'].title()}: {msg['content']}" for msg in messages
+            ])
+            
+            # Create summarization prompt
+            system_prompt = """You are summarizing a D&D campaign conversation for context preservation. Create a concise summary that captures:
```

Comment:
This system prompt has a lot of good elements, but it makes it sound like players will be playing the game by chatting. That is not generally the case. Players will be chatting to remember information, to creatively engage with the existing content, to bounce ideas around, or just to have fun with the content. Keep parts of the prompt that still make sense with that use case, and amend parts that assume the use case is for actively playing the game.